vpr/src/place/place_macro.cpp

```diff
@@ -1,62 +1,144 @@
- * [0...cluster_ctx.clb_nlist.blocks().size()-1]                                    */
-static vtr::vector_map<ClusterBlockId, int> f_imacro_from_iblk;
+/* Go through all the ports in all the blocks to find the port that has the same   *
+ * name as port_name and belongs to the block type that has the name pb_type_name. *
+ * Then, check that whether start_pin_index and end_pin_index are specified. If    *
+ * they are, mark down the pins from start_pin_index to end_pin_index, inclusive.  *
+ * Otherwise, mark down all the pins in that port.                                 */
+static void mark_direct_of_ports(int idirect,
+                                 int direct_type,
+                                 char* pb_type_name,
+                                 char* port_name,
+                                 int end_pin_index,
+                                 int start_pin_index,
+                                 const char* src_string,
+                                 int line,
+                                 std::vector<std::vector<int>>& idirect_from_blk_pin,
+                                 std::vector<std::vector<int>>& direct_type_from_blk_pin);
+
+/* Mark the pin entry in idirect_from_blk_pin with idirect and the pin entry in    *
+ * direct_type_from_blk_pin with direct_type from start_pin_index to               *
+ * end_pin_index.                                                                  */
+static void mark_direct_of_pins(int start_pin_index,
+                                int end_pin_index,
+                                int itype,
+                                int isub_tile,
+                                int iport,
+                                std::vector<std::vector<int>>& idirect_from_blk_pin,
+                                int idirect,
+                                std::vector<std::vector<int>>& direct_type_from_blk_pin,
+                                int direct_type,
+                                int line,
+                                const char* src_string);
+
+const std::vector<t_pl_macro>& PlaceMacros::macros() const {
+    return pl_macros_;
+}
 
-/******************** Subroutine declarations ********************************/
+void PlaceMacros::alloc_and_load_placement_macros(const std::vector<t_direct_inf>& directs) {
+    /* Allocates allocates and loads placement macros and returns
+     * the total number of macros in 2 steps.
+     *   1) Allocate temporary data structure for maximum possible
+     *      size and loops through all the blocks storing the data
+     *      relevant to the carry chains. At the same time, also count
+     *      the amount of memory required for the actual variables.
+     *   2) Allocate the actual variables with the exact amount of
+     *      memory. Then loads the data from the temporary data
+     *       structures before freeing them.
+     *
+     * For pl_macro_member_blk_num, allocate for the first dimension
```

Comment:
Delete comment paragraph from "For pl_macro_member_blk_num ... " to "in the 100k's range)."